network/crl.go

```diff
@@ -0,0 +1,370 @@
+func (c *LRUCRLCache) Put(key string, value *CRLCacheItem) error {
+	c.mutex.Lock()
+	c.cache.Add(key, value)
+	c.mutex.Unlock()
+	log.Debugf("CRL: LRU cache: inserted '%s'", key)
+	return nil
+}
+
+// Remove removes item from cache
+func (c *LRUCRLCache) Remove(key string) error {
+	c.mutex.Lock()
+	c.cache.Remove(key)
+	c.mutex.Unlock()
+	log.Debugf("CRL: LRU cache: removed '%s'", key)
+	return nil
+}
+
+// DefaultCRLVerifier is a default implementation of CRLVerifier
+type DefaultCRLVerifier struct {
+	Config CRLConfig
+	Client CRLClient
+	Cache  CRLCache
+}
+
+// Tries to find cached CRL, fetches using v.Client if not found, checks the signature of CRL using issuerCert
+func (v DefaultCRLVerifier) getCachedOrFetch(uri string, issuerCert *x509.Certificate) (*pkix.CertificateList, error) {
+	// Try v.Cache first, but only if caching is enabled (cache time > 0)
+	if v.Config.cacheTime > 0 {
+		cacheItem, err := v.Cache.Get(v.Config.uri)
+		if cacheItem != nil {
+			if err != nil {
+				// non-empty result + error, should never happen
+				return nil, err
+			}
+
+			if time.Now().Before(cacheItem.Fetched.Add(v.Config.cacheTime)) {
+				return &cacheItem.CRL, nil
+			}
+		}
+	}
+
+	// Not found in cache (or the CRL was outdated), gotta fetch
+	rawCRL, err := v.Client.Fetch(uri)
+	if err != nil {
+		return nil, err
+	}
+
+	crl, err := x509.ParseCRL(rawCRL)
+	if err != nil {
+		log.WithError(err).Debugf("CRL: Cannot parse CRL from '%s'", uri)
+		return nil, err
+	}
+
+	err = issuerCert.CheckCRLSignature(crl)
+	if err != nil {
+		log.WithError(err).Warnf("CRL: Failed to check signature for CRL at %s", uri)
+		return nil, err
+	}
+
+	if crl.TBSCertList.NextUpdate.Before(time.Now()) {
+		log.Warnf("CRL: CRL at %s is outdated", uri)
+		return nil, ErrOutdatedCRL
+	}
+
+	if v.Config.cacheTime > 0 {
+		cacheItem := &CRLCacheItem{Fetched: time.Now(), CRL: *crl}
+		v.Cache.Put(uri, cacheItem)
+	}
+
+	return crl, nil
+}
+
+func (v DefaultCRLVerifier) verifyCertWithIssuer(cert, issuer *x509.Certificate) error {
+	log.Debugf("CRL: Verifying '%s'", cert.Subject.String())
+
+	if len(v.Config.uri) == 0 && v.Config.fromCert == crlFromCertIgnore {
+		log.Debugln("CRL: Skipping check since no config URI specified and we were told to ignore URIs from certificate")
+		return nil
+	}
+
+	for _, crlDistributionPoint := range cert.CRLDistributionPoints {
+		log.Debugf("CRL: certificate contains CRL URI: %s", crlDistributionPoint)
+	}
+
+	queriedCRLs := make(map[string]struct{})
+
+	if v.Config.uri != "" {
+		crl, err := v.getCachedOrFetch(v.Config.uri, issuer)
+		if err != nil {
+			log.WithError(err).Debugf("CRL: Cannot get CRL from '%s'", v.Config.uri)
+			return err
+		}
+
+		for _, revokedCertificate := range crl.TBSCertList.RevokedCertificates {
+			if revokedCertificate.SerialNumber.Cmp(cert.SerialNumber) == 0 {
+				log.Warnf("CRL: Certificate %v was revoked at %v", cert.SerialNumber, revokedCertificate.RevocationTime)
+				return ErrCertWasRevoked
+			}
```

Comment:
Seems like [there is no `Equal()`](https://pkg.go.dev/math/big#Int), only [`Cmp()`](https://pkg.go.dev/math/big#Int.Cmp)

Reply:
I had [this `Equal()`](https://golang.org/pkg/crypto/x509/#Certificate.Equal) in mind 🤔

Reply:
Understood. The problem is, we are not comparing two `x509.Certificate`, but rather [`pkix.RevokedCertificate`](https://pkg.go.dev/crypto/x509/pkix#RevokedCertificate) and [`x509.Certificate`](https://golang.org/pkg/crypto/x509/#Certificate). CRLs do not contain whole certificates, but rather serial number and (optionally) some extensions. So I'll leave `Cmp()` as is. Also, seems like I missed that `Extensions` handling, while [RFC tells](https://tools.ietf.org/html/rfc5280#section-4.2) that we must process it, and deny certificate in case of any _critical_ extension cannot be processed (i.e. we dunno how to handle it).